stage0/README.md

```diff
@@ -37,3 +63,113 @@ To use the binary, pass it to `qemu -bios`; for example:
 ```shell
 qemu-system-x86_64 -nodefaults -nographic -no-reboot -machine microvm -bios stage0.bin
 ```
+
+## Loading the kernel
+
+There are two options for loading the kernel: either get the VMM to load it into
+guest memory before the VM even starts, or let stage0 load the kernel from
+`fw_cfg`. Both approaches have their benefits and drawbacks.
+
+**Note**: the kernel has to be an ELF file! If you want to use Linux, use the
+`extract-vmlinux` script to create the uncompressed `vmlinux` image from the
+`bzImage` file.
+
+### Pre-loading the kernel
+
+You can use the QEMU `loader` device to load a kernel into memory:
+
+```shell
+qemu-system-x86_64 [...] -device loader,file=/path/to/kernel
+```
+
+As long as the file is in ELF format, QEMU will lay out the binary in guest
+memory based on the instructions in the ELF file.
+
+Notes:
+
+- The `loader` device does not ask the PSP to encrypt the memory under SEV, so
+  this will only work with no memory encryption enabled.
+- stage0 will assume that there is an ELF header at address `0x20_0000` (2 MiB).
+  If no ELF header is found at that address, stage0 will just blindly jump to
+  that address, expecting it to be executable.
+
+### Loading the kernel via `fw_cfg`
+
+We also support loading the kernel (and an initial ramdisk) via two custom
+`fw_cfg` entries:
+
+```shell
+qemu-system-x86_64 [...] -fw_cfg name=opt/stage0/elf_kernel,file=/path/to/kernel \
+                         -fw_cfg name=opt/stage0/initramfs,file=/path/to/initramfs \
+                         -fw_cfg name=opt/stage0/cmdline,string=console=ttyS0
+```
+
+Unfortunately we had to implement custom `fw_cfg` entries; the standard
+`-kernel` flag won't work as QEMU may load files into guest memory, but it will
+not ask the PSP to encrypt the memory. If we don't provide a `-kernel` flag,
+QEMU will reject `-initrd` and `-append` flags; thus, we had to reimplement
+those as well.
+
+This approach is supported under SEV and SEV-ES. (Probably also SEV-SNP, but the
+public releases of QEMU do not support SEV-SNP yet.)
+
+## Internals
+
+### Memory layout
+
+Stage0 maps the first 1 GiB of physical memory using identity mapping before
+handing control over to the kernel. The stage0 binary itself is mapped into
+memory just below the end of 4 GiB address space, like any other BIOS ROM.
+
+```text
+￼              |                      ...                       |
```

Comment:
Not sure what these extra character are, or whether they will show up when the markdown is rendered.

Reply:
Which characters are you referring to?

Reply:
![image](https://user-images.githubusercontent.com/387628/232881124-bfff4a6c-3d44-4464-b16b-66e7c818d99e.png)

Reply:
I couldn't see those mysterious characters, but thanks to @tiziano88's help they should all be gone.